src/charm.py

```diff
@@ -1,104 +1,402 @@
+                self._pgbouncer_container: {
+                    "summary": "pgbouncer service",
+                    "user": self._pgbouncer_user,
+                    "command": f"pgbouncer {INI_PATH}",
                     "startup": "enabled",
-                    "environment": {"thing": self.model.config["thing"]},
+                    "override": "replace",
+                    "environment": {
+                        "PGB_DATABASES": self.config["pgb_databases"],
+                        "PGB_LISTEN_PORT": self.config["pgb_listen_port"],
+                        "PGB_LISTEN_ADDRESS": self.config["pgb_listen_address"],
+                        "PGB_ADMIN_USERS": self.config["pgb_admin_users"],
+                    },
                 }
             },
         }
-        # Add intial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", pebble_layer, combine=True)
-        # Autostart any services that were defined with startup: enabled
+
+    def _on_pgbouncer_pebble_ready(self, event: PebbleReadyEvent) -> None:
+        """Define and start pgbouncer workload."""
+        container = event.workload
+        pebble_layer = self._pgbouncer_layer()
+        container.add_layer(self._pgbouncer_container, pebble_layer, combine=True)
         container.autostart()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
         self.unit.status = ActiveStatus()
 
-    def _on_config_changed(self, _):
-        """Just an example to show how to deal with changed configuration.
+    # ====================
+    #  Charm Action Hooks
+    # ====================
+
+    def _on_reload_pgbouncer_action(self, event: ActionEvent) -> None:
+        """An action to explicitly reload pgbouncer application.
+
+        It should be made obvious to the user that this does not restart the container nor the
+        charm - it *only* restarts the application itself.
+        """
+        self._reload_pgbouncer()
+        event.set_results({"result": "pgbouncer application has restarted"})
+
+    def _on_change_password_action(self, event: ActionEvent) -> None:
+        """An action to update the password for a specific user.
+
+        Currently passwords are hashed using md5.
+
+        Args:
+            event: ActionEvent containing a "username" parameter, defining the user whose password
+                will be modified, and a "password" parameter, defining the new password.
+        """
+        username = event.params["username"]
+        users = self._get_userlist_from_container()
+        if username not in users:
+            event.set_results(
+                {
+                    "result": f"user {username} does not exist - use the get-users action to list existing users."
+                }
+            )
+            return
+        password = hashlib.md5(event.params["password"].encode())
```

Comment:
comment

Reply:
I've extracted hashing into a separate function, which is independently documented. Does this adequately explain the function of the hashing, or shall I elaborate?